crates/server/src/server/h3_handler.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2015-2021 Benjamin Fry <benjaminfry@me.com>
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+use std::{io, net::SocketAddr, sync::Arc};
+
+use bytes::{Buf, Bytes};
+use drain::Watch;
+use futures_util::lock::Mutex;
+use h3::server::RequestStream;
+use h3_quinn::BidiStream;
+use tracing::{debug, warn};
+use trust_dns_proto::{
+    error::ProtoError, h3::h3_server::H3Connection, http::Version, https::HttpsError, rr::Record,
+};
+
+use crate::{
+    authority::MessageResponse,
+    server::{
+        request_handler::RequestHandler, response_handler::ResponseHandler, server_future,
+        Protocol, ResponseInfo,
+    },
+};
+
+pub(crate) async fn h3_handler<T>(
+    handler: Arc<T>,
+    mut connection: H3Connection,
+    src_addr: SocketAddr,
+    _dns_hostname: Option<Arc<str>>,
+    shutdown: Watch,
+) -> Result<(), ProtoError>
+where
+    T: RequestHandler,
+{
+    // TODO: we should make this configurable
+    let mut max_requests = 100u32;
+
+    // Accept all inbound requests sent over the connection.
+    loop {
+        let (_, mut stream) = tokio::select! {
+            result = connection.accept() => match result {
+                Some(Ok(next_request)) => next_request,
+                Some(Err(err)) => {
+                    warn!("error accepting request {}: {}", src_addr, err);
+                    return Err(err);
+                }
+                None => {
+                    break;
+                }
+            },
+            _ = shutdown.clone().signaled() => {
+                // A graceful shutdown was initiated.
+                break;
+            },
+        };
+
+        let request = match stream
+            .recv_data()
+            .await
+            .map_err(|e| ProtoError::from(format!("h3 stream receive data failed: {e}")))?
+        {
+            Some(mut request) => request.copy_to_bytes(request.remaining()),
+            None => {
+                warn!("h3 stream received no data");
+                continue;
+            }
+        };
+
+        debug!(
+            "Received bytes {} from {src_addr} {request:?}",
+            request.remaining()
+        );
+        let handler = handler.clone();
+        let stream = Arc::new(Mutex::new(stream));
+        let responder = H3ResponseHandle(stream.clone());
+
+        handle_request(request, src_addr, handler, responder).await;
+
+        max_requests -= 1;
+        if max_requests == 0 {
+            warn!("exceeded request count, shutting down h3 conn: {src_addr}");
```

Comment:
Same here, I would have said this should return an error instead of just a warning.

Reply:
I'm trying to remember my exact reasoning on the return of Err or not here. I think it was that shutting down the connection is not an error condition, just normal behavior. But we could change that logic...

Reply:
No, that makes sense to me actually.
I guess it would be nice to notify users if a protection mechanism was triggered, but that has to be changed in the whole library, not just here.